jdaviz/core/region_translators.py

```diff
@@ -0,0 +1,226 @@
+"""The ``region_translators`` module houses translations of
+:ref:`regions:shapes` to :ref:`photutils:photutils-aperture` apertures.
+"""
+from astropy import units as u
+from astropy.coordinates import Angle
+from photutils.aperture import (CircularAperture, SkyCircularAperture,
+                                EllipticalAperture, SkyEllipticalAperture,
+                                RectangularAperture, SkyRectangularAperture,
+                                CircularAnnulus, SkyCircularAnnulus,
+                                EllipticalAnnulus, SkyEllipticalAnnulus,
+                                RectangularAnnulus, SkyRectangularAnnulus)
+from regions import (CirclePixelRegion, CircleSkyRegion,
+                     EllipsePixelRegion, EllipseSkyRegion,
+                     RectanglePixelRegion, RectangleSkyRegion,
+                     CircleAnnulusPixelRegion, CircleAnnulusSkyRegion,
+                     EllipseAnnulusPixelRegion, EllipseAnnulusSkyRegion,
+                     RectangleAnnulusPixelRegion, RectangleAnnulusSkyRegion, PixCoord)
+
+__all__ = ['regions2aperture', 'aperture2regions']
+
+
+def regions2aperture(region_shape):
+    """Convert a given ``regions`` shape to ``photutils`` aperture.
+
+    Parameters
+    ----------
+    region_shape : `regions.Region`
+        A supported ``regions`` shape.
+
+    Returns
+    -------
+    aperture : `photutils.aperture.Aperture`
+        An equivalent ``photutils`` aperture.
+
+    Raises
+    ------
+    NotImplementedError
+        The given ``regions`` shape is not supported.
+
+    Examples
+    --------
+    Translate a `regions.CirclePixelRegion` to `photutils.aperture.CircularAperture`:
+
+    >>> from regions import CirclePixelRegion, PixCoord
+    >>> from jdaviz.core.region_translators import regions2aperture
+    >>> region_shape = CirclePixelRegion(center=PixCoord(x=42, y=43), radius=4.2)
+    >>> regions2aperture(region_shape)
+    <CircularAperture([42., 43.], r=4.2)>
+
+    Also see
+    --------
+    aperture2regions
+
+    """
+    if isinstance(region_shape, CirclePixelRegion):
+        aperture = CircularAperture(region_shape.center.xy, region_shape.radius)

```

Comment:
The aperture parameters will be references to the region parameters.  It's probably better to copy the region parameters when initializing the aperture shapes.  Otherwise, if someone later changes a region parameter the aperture will unexpectedly change.

Reply:
@larrybradley ... do you think it is better to have the copy done inside photutils/regions? Like how one can choose to copy for not for Quantity? I will do the copy here for now but something to think about.

Reply:
I'm not sure what you mean.  It's not standard to have functions that immediately copy of the their inputs. `Quantity`, `SkyCoord`, and `Angle` all have a copy (deepcopy) method.  `Region` objects also have a copy method.  `Aperture` objects do not, but I'm adding that right now!  It'll be easier to just copy the region/aperture than copy each individual parameter.

Reply:
You don't think maybe some users want to keep the reference? I mean in Quantity, there is a `copy=` option to allow that flexibility. https://docs.astropy.org/en/stable/units/quantity.html (see the Note)

Reply:
Actually... this might only be a problem for sky shapes/apertures, for which I pass in the Quantity directly. For non-Quantity scalar, I don't think reference sticks.

```python
>>> from photutils.aperture import CircularAperture
>>> from regions import CirclePixelRegion, PixCoord
>>> region_shape = CirclePixelRegion(center=PixCoord(x=42, y=43), radius=4.2)
>>> aperture = CircularAperture(region_shape.center.xy, region_shape.radius)
>>> aperture
<CircularAperture([42., 43.], r=4.2)>
>>> region_shape.center.x = 55
>>> region_shape
<CirclePixelRegion(center=PixCoord(x=55, y=43), radius=4.2)>
>>> aperture
<CircularAperture([42., 43.], r=4.2)>
```

Reply:
Also, I don't think it is actually passed by reference. See below. But if it makes you feel better, I can do a `deepcopy` of the input in the conversion function. What do you think?

```python
>>> region_sky  # From the test suite
<CircleSkyRegion(center=<SkyCoord (ICRS): (ra, dec) in deg
    (337.50779898, -20.82138816)>, radius=4.199999726593498 arcsec)>
>>> aperture_sky = SkyCircularAperture(region_sky.center, region_sky.radius)
>>> aperture_sky
<SkyCircularAperture(<SkyCoord (ICRS): (ra, dec) in deg
    (337.50779898, -20.82138816)>, r=4.199999726593498 arcsec)>
>>> region_sky.radius = 1 * u.arcsec
>>> region_sky
<CircleSkyRegion(center=<SkyCoord (ICRS): (ra, dec) in deg
    (337.50779898, -20.82138816)>, radius=1.0 arcsec)>
>>> aperture_sky  # Still the old radius
<SkyCircularAperture(<SkyCoord (ICRS): (ra, dec) in deg
    (337.50779898, -20.82138816)>, r=4.199999726593498 arcsec)>
```

On the other hand, changing `SkyCircularAperture` parameter directly has no effect even on itself:

```python
>>> aperture_sky.radius = 1 * u.arcsec
>>> aperture_sky
<SkyCircularAperture(<SkyCoord (ICRS): (ra, dec) in deg
    (337.50779898, -20.82138816)>, r=4.199999726593498 arcsec)>
```

Reply:
Ah, OK.  I think this is because these parameters are actually descriptors.